server/gokbg3/grails-app/services/org/gokb/ValidationService.groovy

```diff
@@ -0,0 +1,610 @@
+
+    int ctr = 0
+
+    header.each {
+      col_positions[it] = ctr++
+    }
+
+    knownColumns.each { colName, info ->
+      if (info.mandatory && !header.contains(colName)) {
+        result.errors.missingColumns.add(colName)
+      }
+      else if (!header.contains(colName) && colName != 'zdb_id') {
+        result.warnings.missingColumns.add(colName)
+      }
+    }
+
+    String[] nl = csv.readNext()
+    int rowCount = 0
+
+    if (result.errors.missingColumns.size() == 0) {
+      while (nl != null) {
+        rowCount++
+
+        if (nl.size() > 1 && nl.size() != header.size()) {
+          result.rows.total++
+          result.rows.error++
+          result.errors.rows["${rowCount}"] = [
+            columnsCount: [
+              message: "Inconsistent column count (${nl.size()} <> ${header.size()})!",
+              messageCode: "kbart.errors.tabsCountRow",
+              args: [nl.size(), header.size()]
+            ]
+          ]
+          if (!result.errors.type['columnCount']) {
+            result.errors.type['columnCount'] = 1
+          }
+          else {
+            result.errors.type['columnCount']++
+          }
+          result.valid = false
+        }
+        else if (nl.size() >= 5) {
+          result.rows.total++
+
+          def row_result = checkRow(nl, rowCount, col_positions, titleIdNamespace)
+
+          if (row_result.errors) {
+            result.rows.error++
+            result.valid = false
+            result.errors.rows["${rowCount}"] = row_result.errors
+
+            row_result.errors.each { error_key, error_list ->
+              if (!result.errors.type[error_key]) {
+                result.errors.type[error_key] = 1
+              }
+              else {
+                result.errors.type[error_key]++
+              }
+            }
+          }
+          if (row_result.warnings) {
+            result.rows.warning++
+            result.warnings.rows["${rowCount}"] = row_result.warnings
+
+            row_result.warnings.each { warn_key, warn_list ->
+              if (!result.warnings.type[warn_key]) {
+                result.warnings.type[warn_key] = 1
+              }
+              else {
+                result.warnings.type[warn_key]++
+              }
+            }
+          }
+        }
+        else {
+          log.debug("Found and skipped short row ${nl}")
+          result.rows.warning++
+
+          if (!result.warnings.rows["${rowCount}"]) {
+            result.warnings.rows["${rowCount}"] = [:]
+          }
+
+          result.warnings.rows["${rowCount}"]["shortRow"] = [
+            message: "Skipped short/empty row!",
+            messageCode: "kbart.errors.shortRow"
+          ]
+        }
+
+        nl = csv.readNext()
+      }
+    }
+    else {
+      log.debug("Missing mandatory columns... skipping file processing!")
+    }
+
+    csv.close()
+
+    result
+  }
+
+  private CSVReader initReader (the_data) {
+    final CSVParser parser = new CSVParserBuilder()
```

Comment:
A byte order mark (BOM) would break Ygor's processing, unless we had included BOM filtering. I guess we'd need that here as well?